### PR TITLE
`matrix_exp`: Make sure `_compute_linear_combinations` result preserves dim of the input.

### LINKED ISSUE #80948: Weird warning using matrix exp for complex tensors
### 🐛 Describe the bug

Warning appears if you run this snippet
```python
torch.manual_seed(42)
tens = 0.5 * torch.randn(10, 3, 3, dtype=torch.complex64)
tens = (0.5 * (tens.transpose(-1, -2) + tens))
tens.imag = torch.matrix_exp(tens.imag)
```
The warning
```
/tmp/ipykernel_18699/1530108168.py:4: UserWarning: An output with one or more elements was resized since it had shape [4, 3, 3], which does not match the required output shape [1, 4, 3, 3]. This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at  ../aten/src/ATen/native/Resize.cpp:24.)
  tens.imag = torch.matrix_exp(tens.imag)
```

It did not happen prior to 1.12 version

### Versions

Collecting environment information...
PyTorch version: 1.12.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.1.0
Clang version: 13.0.1
CMake version: version 3.23.1
Libc version: glibc-2.35

Python version: 3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:24:37)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.17.4-arch1-1-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.112
GPU models and configuration: 
GPU 0: NVIDIA GeForce GTX 1070
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090

Nvidia driver version: 510.60.02
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.3.3
/usr/lib/libcudnn_adv_infer.so.8.3.3
/usr/lib/libcudnn_adv_train.so.8.3.3
/usr/lib/libcudnn_cnn_infer.so.8.3.3
/usr/lib/libcudnn_cnn_train.so.8.3.3
/usr/lib/libcudnn_ops_infer.so.8.3.3
/usr/lib/libcudnn_ops_train.so.8.3.3
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.0
[pip3] torch==1.12.0
[conda] numpy                     1.22.0                   pypi_0    pypi
[conda] torch                     1.12.0                   pypi_0    pypi

cc @ezyang @gchanan @zou3519 @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @jianyuh @pearu @walterddr @IvanYashchuk @xwang233

### PR DESCRIPTION
**Pull request by nikitaved**
Fixes https://github.com/pytorch/pytorch/issues/80948.

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
There was the pesky `squeeze` that was missing. Thank you @nikitaved!

Could you add a short test to make sure that we're fixing the issue?

**Comment by Lezcano** [comment]
Yup, could you add one of those wrappers that make the test fail if a warning is thrown around those operations?

**Comment by nikitaved** [comment]
I seem fail to capture this warning with OpInfo, for example:
```
================================================================================= FAILURES ==================================================================================
_______________________________________________________________ TestCommonCPU.test_out_warning_matrix_exp_cpu _______________________________________________________________
Unexpected success
______________________________________________________________ TestCommonCUDA.test_out_warning_matrix_exp_cuda ______________________________________________________________
Unexpected success
============================================================================= warnings summary ==============================================================================
...
  /home/nik/git/Quansight/pytorch/torch/testing/_internal/common_methods_invocations.py:953: UserWarning: An output with one or more elements was resized since it had shape [1, 3, 3], which does not match the required output shape [1, 1, 3, 3]. This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at /home/nik/git/Quansight/pytorch/aten/src/ATen/native/Resize.cpp:17.)
    return self.op(*args, **kwargs)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================================== short test summary info ==========================================================================
FAILED test/test_ops.py::TestCommonCPU::test_out_warning_matrix_exp_cpu
FAILED test/test_ops.py::TestCommonCUDA::test_out_warning_matrix_exp_cuda
```
This is done with `expectedFailure`. Am I missing something, @mruberry ?

**Comment by Lezcano** [comment]
`test_out_warning` just tests for some warnings, like:
```
            # Case Zero: out= with the correct dtype and device, but the wrong shape
            #   Expected behavior: if nonempty, resize with a warning.
```

What I was suggesting was to add a short test in `test_linalg.py` using something like https://stackoverflow.com/questions/45671803/how-to-use-pytest-to-assert-no-warning-is-raised. Doing so for the matrix example in the issue would be enough.

File: test/test_linalg.py [hunk 1]
```diff
@@ -6189,6 +6190,19 @@ def run_test(coeff_shape, data_shape):
         run_test([3, 4], [3, 3, 3])
         run_test([3, 4], [3, 3, 3, 3])
 
+    @onlyCPU
+    @skipCPUIfNoLapack
+    @dtypes(torch.complex64)
+    def test_linalg_matrix_exp_no_warnings(self, device, dtype):
+        # this tests https://github.com/pytorch/pytorch/issues/80948
+        with freeze_rng_state():
+            torch.manual_seed(42)
```
Comment (nikitaved) [inline test/test_linalg.py:6199]: Not sure this is strictly necessary, but decided to make it play it extra safe.

**Review by lezcano — APPROVED** [review]
Good stuff. Thank you Nik!

**Comment by Lezcano** [comment]
No one can say I didn't try. cc @mruberry

**Review by mruberry — APPROVED** [review]
Stamped!

**Pull request closed without merging**